In [60]:
import numpy as np
import pandas as pd

In [61]:
df = pd.read_csv("data/loan_data.csv")

In [62]:
df.drop(columns = ['Unnamed:0'] , inplace=True , errors = 'ignore')

In [63]:
print('before cleaning :',df.shape)

before cleaning : (150000, 12)


In [64]:
print(df.columns.tolist())

['Unnamed: 0', 'SeriousDlqin2yrs', 'RevolvingUtilizationOfUnsecuredLines', 'age', 'NumberOfTime30-59DaysPastDueNotWorse', 'DebtRatio', 'MonthlyIncome', 'NumberOfOpenCreditLinesAndLoans', 'NumberOfTimes90DaysLate', 'NumberRealEstateLoansOrLines', 'NumberOfTime60-89DaysPastDueNotWorse', 'NumberOfDependents']


In [65]:
df = df.drop(columns=["Unnamed: 0"])
print(df.shape)
print(df.columns.tolist())

(150000, 11)
['SeriousDlqin2yrs', 'RevolvingUtilizationOfUnsecuredLines', 'age', 'NumberOfTime30-59DaysPastDueNotWorse', 'DebtRatio', 'MonthlyIncome', 'NumberOfOpenCreditLinesAndLoans', 'NumberOfTimes90DaysLate', 'NumberRealEstateLoansOrLines', 'NumberOfTime60-89DaysPastDueNotWorse', 'NumberOfDependents']


TASK1 - REMOVE BAD AGE ROWS

In [66]:
df = df[(df['age'] > 0) & (df['age'] <= 100)]
print('after removing the bad age rows:',df.shape)

after removing the bad age rows: (149986, 11)


df["age"] > 0       → keep rows where age is more than 0
df["age"] <= 100    → keep rows where age is 100 or less
&                   → BOTH conditions must be true
df = df[...]        → overwrite df with only the rows that passed

TASK2 - FILL MISSING MONTHLY INCOME

In [74]:
median_income = df['MonthlyIncome'].median()
print("median_income:",median_income)

df['MonthlyIncome'] = df['MonthlyIncome'].fillna(median_income)
print("Missing values in MonthlyIncome:", df["MonthlyIncome"].isnull().sum())

median_income: 5400.0
Missing values in MonthlyIncome: 0


df["MonthlyIncome"].median()   → calculate the middle value of income column
.fillna(median_income)         → wherever there is a NaN, put median_income there
df["MonthlyIncome"] = ...      → save the result back into the same column
.isnull().sum()                → verify — should now show 0 missing

TASK3 - FILL MISSING NUMBER OF DEPENDENTS

In [75]:
median_dependants = df['NumberOfDependents'].median()
print('median_dependants',median_dependants)

df['NumberOfDependents'] = df['NumberOfDependents'].fillna(median_dependants)
print('Missing values in no. of dependants:',df['NumberOfDependents'].isnull().sum())

median_dependants 0.0
Missing values in no. of dependants: 0


.median()        → finds middle value of dependents column
.fillna(...)     → fills all empty cells with that median value
.isnull().sum()  → should show 0 after filling

Median dependents is 0 — meaning more than half the borrowers in this dataset have no dependents.

TASK4 - HANDLE INCOME OUTLIERS

In [82]:
df['MonthlyIncome'] = df['MonthlyIncome'].clip(upper=50000)

print('max value after clipping',df['MonthlyIncome'].max())
print('mean value after clipping',df['MonthlyIncome'].mean().round(2))

max value after clipping 50000.0
mean value after clipping 6260.39


.clip(upper=50000)   → any value above 50000 becomes 50000
                       values below 50000 stay exactly as they are

TASK 5 - VERIFY EVERYTHING IS CLEAN

print('Final shape:',  df.shape)
print('Missing values:',df.isnull().sum())
print('\nAge min',  df['age'].min())
print('Age max',  df['age'].max())
print('max monthly income',  df['MonthlyIncome'].max())

In [ ]:
TASK6 - SAVE THE CLEAN DATA

In [84]:
df.to_csv("data/loan_data_cleaned.csv",index=False)
print('clean data saved successfully!')

clean data saved successfully!


df.to_csv(...)        → saves df as a CSV file
"data/loan_..."       → saves it inside your data folder
index=False           → don't save the row numbers as a column
                        remember the Unnamed: 0 problem? this prevents that

 Interviews-

"After EDA I performed data cleaning — I removed 14 records with invalid age values, used median imputation for MonthlyIncome which had 19.82% missing values, and capped income outliers at $50,000 since the raw maximum of $3 million was clearly unrealistic. After cleaning the dataset had 149,986 records with zero missing values."